### ISSUE TITLE
Embedding jupyterhub in iframe

### ISSUE THREAD

**Issue by satendrakumar06** [jupyter/jupyterhub#379]
I have embedded jupyter notebook inside iframe, It's working fine. Now, I am trying to embed Jupyterhub inside iframe.  But, getting error  "Blocked by Content Security Policy". Is Any way to do this?

**Comment by minrk** [comment]
@rgbkrk do you remember any gotchas for this one?

**Comment by satendrakumar06** [comment]
@minrk @rgbkrk I have tried configuration setting:
```
c.JupyterHub.tornado_settings = { 'headers': { 'Content-Security-Policy': "frame-ancestors 'self' http://122.177.83.47:9000/#/analytics/notebook" } }
```

It didn't  work. Same thing i have  tried in notebook it is working fine.

**Comment by rgbkrk** [comment]
The single user notebook servers will also need the Content Security Policy set.

**Comment by rgbkrk** [comment]
As in the notebook config for each server will need this setting:

```python
c.NotebookApp.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'self' 'ec2-url:9000/analytics/notebook' "
  }
}
```

**Comment by satendrakumar06** [comment]
@rgbkrk Where, I will set this config ?

**Comment by rgbkrk** [comment]
How are you launching the single user servers? Is this the default JupyterHub setup or are you using one of the customer spawners?

**Comment by satendrakumar06** [comment]
This is the default JupyterHub setup.

**Comment by charleswardell** [comment]
Hi JupyterHUB team, our development team is really stuck and was wondering if there is someway you can help or point us to perhaps a consultant that we can hire that could help us troubleshoot the issue that @satendrakumar06 posted above. We are getting error "Blocked by Content Security Policy" when embedding JupyterHub in an IFRAME. Any help you can provide us would be greatly appreciated. Thank you.

**Comment by jerowe** [comment]
+1 for this. We are trying to get jupyterhub in an iframe. We tried the

'Content-Security-Policy': "frame-ancestors 'self' 'ec2-url:9000/analytics/notebook' "

Maybe I don't have the url right? Jupyterhub is being served at port 8000, with a base url of ipython.

**Comment by charleswardell** [comment]
@jerowe did you have any luck on this? I am not sure or posts are going noticed.

**Comment by minrk** [comment]
Sorry for the slow response. I tested this myself a bit, and I discovered a few things:

1. I don't think a path should be in the 
2. when I test, I cannot seem to get both 'self' *and* an alternate host to work.

So while the following doesn't work:

    'Content-Security-Policy': "frame-ancestors 'self' 'ec2-url:9000/analytics/notebook' "

This does:

    'Content-Security-Policy': "frame-ancestors 'ec2-url:9000' "

Full config:

```python
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'ec2-url:9000'",
    }
}
```

Note that you will have to set this for *both* the Hub itself and the single-user servers (in `/etc/jupyter/jupyter_notebook_config.py`)

**Comment by satendrakumar06** [comment]
@minrk Thanks!!! It is working fine.

**Comment by minrk** [comment]
Thanks for confirming that it works!

I can't understand *why* `'self' 'host'`doesn't work, since the `src-list` spec seems to suggest that it should, but I've at least tested both `'self' 'host'` and `'host' 'self'`, and in both cases the first value was used and the second ignored, in both Chrome and Firefox. Maybe it's a bug, and maybe it's my misunderstanding of the spec.

**Comment by jerowe** [comment]
No joy for me. I'm still getting a Blocked by Content Security Policy.

**Comment by shusson** [comment]
Tested and working with 7.2

jupyter_notebook_config.py:
```
c.NotebookApp.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'self' http://parenturl.com "
    }
}
```

jupyterhub_config.py:
```
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'self' http://parenturl.com"
  }
}
```

**Comment by raemounz** [comment]
May I know which location to edit the jupyter_notebook_config.py? I was able to embed the login in iFrame (after editing jupyterhub_config.py) but after user logs in, I got the content-security-policy issue.  I edited the jupyter_notebook_config.py inside /root/.jupyter.

**Comment by shusson** [comment]
It depends on how you run your notebook server, by default it's `~/.jupyter`. See http://jupyter.readthedocs.io/en/latest/projects/jupyter-directories.html for more info

**Comment by gaorongchao** [comment]
@shusson ,Thank you very much ,your setting is work for me !

**Comment by TapasSpark** [comment]
I am trying to embed the jupyterhub in an iframe in a portal. After logging into the jupyterhub I am getting the error :
Refused to display **'http://host_ip:8000/user/user_name/tree?redirects=1' in a frame because an ancestor violates the following Content Security Policy directive: "frame-ancestors 'self'".**

I have put the below tornado settings in notebook as well as jupyterhub .py files
jupyter_notebook_config.py:
c.NotebookApp.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'self' "
    }
}
jupyterhub_config.py:
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'self' "
  }
}

The same setting was working for jupyter notebook but login was not there.

Any clue for this ?

**Comment by satendrakumar** [comment]
Try this: 
```
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'ec2-url:9000'",
    }
}
```

**Comment by TapasSpark** [comment]
Hi satendrakumar,
with the suggested settings the login page itself is not getting loaded into iframe. I tried with below type of tornado settings for jupyterhub.
**1)**
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'http://host_ip:8000'",
    }
}
**2)**
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'host_ip:8000'",
    }
}

**Comment by TapasSpark** [comment]
@satendrakumar ,
With the below settings the login page is getting embedded in iframe but after login it becomes blank and the error appears in console: **Refused to display 'http://host_ip:8000/user/user_name/tree?redirects=1' in a frame because an ancestor violates the following Content Security Policy directive: "frame-ancestors 'self'".**

c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors * 'self' "
    }
}

**Comment by satendrakumar** [comment]
Have you tried this:
```
c.JupyterHub.tornado_settings = {
'headers': {
'Content-Security-Policy': "frame-ancestors 'host_ip:8000'",
}
}
```

**Comment by TapasSpark** [comment]
@satendrakumar,
Yes I had tried that one too.

**Comment by satendrakumar** [comment]
It is working for me. @minrk Any suggestion ?

**Comment by fbessou** [comment]
@TapasSpark,
in order for this configuration to work, the host hostname:ip should not be single quoted.

So use
```
'Content-Security-Policy': "frame-ancestors host_ip:8000 ;",
```
instead of
```
'Content-Security-Policy': "frame-ancestors 'host_ip:8000'",
```

**Comment by collinwo** [comment]
I got the same problem with @TapasSpark . By using

 'Content-Security-Policy': "frame-ancestors host_ip:8000 ;",

Error
Refused to display 'http://ABC:8713/user/yucol/tree?' in a frame because an ancestor violates the following Content Security Policy directive: "frame-ancestors 'self'".

Any suggestions?

**Comment by arkroop** [comment]
Hi @minrk,

I have installed jupyterhub on Linux machine on EC2 within user name ec2-user. In the jupyterhub configuration I allowed to add user as a Linux machine user when we add user in jupyterhub using API. After adding the user I created the new notebook and generate the token of the same user. I used Iframe to show the notebook with another user but getting the content security policy issue. I tried to all suggested solution but not success please provide the solution of this issue.

**Comment by zcesur** [comment]
I had the same issue and came up with [this](https://github.com/zcesur/embeddable-jhub-sample) configuration after reading the answers above. Might be helpful for those of you trying to run JupyterHub on Kubernetes.

**Comment by kouohhashi** [comment]
Hi @satendrakumar ,

I have same issue.
Have you find a solution?

embedding https://dmain/hub/login is okay
but https://dmain/user/user_name/tree?redirects=1 got error, "an ancestor violates the following Content Security Policy directive: "frame-ancestors 'self'".

My jupyterhub_config.py setting file: 

```
c.JupyterHub.tornado_settings={
  'headers':{
    'Content-Security-Policy': "frame-ancestors *;"
  }
}

c.NotebookApp.tornado_settings={
  'headers':{
    'Content-Security-Policy': "frame-ancestors *;"
  }
}
```

I'm guessing I need to edit some place other than jupyterhub_config.py because my config file even don't have "self".

**Comment by satendrakumar** [comment]
@kouohhashi [It's working fine for me.](https://github.com/jupyterhub/jupyterhub/issues/379#issuecomment-176639586)

**Comment by kouohhashi** [comment]
Thanks, I figured out my problem.

It turned out I needed to set  **_c.NotebookApp.tornado_settings_** in docker images' jupyter_notebook_config.py. 

And looks like avoiding double quotation marks make src-list work.

**Worked**
```
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': 'frame-ancestors self http://localhost:3000',
    }
}
```

**Not worked. 2nd src was not taken**
```
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors 'self' 'http://localhost:3000'",
    }
}
```

**Comment by TapasSpark** [comment]
Hi ,
I have embedded jupyterhub with LDAP authentication in iframe of my web application. But eeverytime jupyterhub loads the login page is prompted. I want the users should login seamlessly without the login page presented to them and simultaneously the LDAP authetication of jupyterhub should happen. Is this achievable ? If somebody has already done this, please suggest some ways to do this.

Thanks

**Comment by marcomoretto** [comment]
> Hi ,
> I have embedded jupyterhub with LDAP authentication in iframe of my web application. But eeverytime jupyterhub loads the login page is prompted. I want the users should login seamlessly without the login page presented to them and simultaneously the LDAP authetication of jupyterhub should happen. Is this achievable ? If somebody has already done this, please suggest some ways to do this.
> 
> Thanks

Hi @TapasSpark ,
I solved it with REST API and token login https://jupyterhub.readthedocs.io/en/stable/reference/rest.html
When a user (already logged in my system) request for a jupyterhub session, I request a token (using an admin token I defined in the jupyterhub_config file), something like this (I did it in Python but you'll have an idea):
```
su_token = 'xxx' # my secret token defined in jupyterhub_config for my admin
r = requests.post(api_url + '/users/' + request.user.username + '/tokens',
                          headers={
                              'Authorization': 'token %s' % su_token
                          }
                          )
r.raise_for_status()
token = r.json()['token']
```

then I start the server for the user 

```
r = requests.post(api_url + '/users/' + request.user.username + '/server',
                          headers={
                              'Authorization': 'token %s' % su_token
                          }
                          )
r.raise_for_status()

```

Finally I return the URL for the user (this should show the home directory for that user)

`base_jhub_ip += '/user/' + request.user.username + '/tree/?token=' + token`

**Comment by meesam15** [comment]
None of these suggestions worked work for me except for this one: 
'Content-Security-Policy': "frame-ancestors * 'self' "

**Comment by 2efPer** [comment]
It's not working when someone is login.
Refused to display 'http://hostname:8000/user/USERNAME/tree?redirects=1' in a frame because an ancestor violates the following Content Security Policy directive: "frame-ancestors 'self'".

my configuration is : 

> 'Content-Security-Policy': "frame-ancestors * 'self' http://0.0.0.0:8000" .

**Comment by 2efPer** [comment]
all of this does not work for me. 
My environment is created by virtualenv, ` c.NotebookApp.tornado_settings ` has no effect on this problem when I add this to jupyterhub_config.py. Where can I set this attribute since there is no **jupyter_notebook_config.py** ?
or can I change pass by ` c.Spawner.args `?

**Comment by vladmihaisima** [comment]
@2efPer @djknight1 For me putting jupyter_notebook_config.py at /etc/jupyter/jupyter_notebook_config.py worked. To test maybe you can use "frame-ancestors *;" (might be easier to check if it works at all)

**Comment by djknight1** [comment]
@vladmihaisima Thanks for replying.It works only if using jupyter notebook，it failed when i use jupyterhub.

**Comment by vladmihaisima** [comment]
@djknight1  I have used the docker image at https://hub.docker.com/r/jupyterhub/jupyterhub/, changed both the jupyterhub config (at /srv/jupyterhub/jupyterhub_config.py) and the jupyter notebooks config (at  /etc/jupyter/jupyter_notebook_config.py), without any other change (local spawner)

**Comment by djknight1** [comment]
@vladmihaisima Thanks.I will try.

**Comment by 2efPer** [comment]
I found out the easiest way to config, it should work in almost all situation.
add this to your `jupyterhub_config.py`: 
```
c.Spawner.args = ['--NotebookApp.tornado_settings={"headers":{"Content-Security-Policy": "frame-ancestors * self host_ip:port"}}']   

c.JupyterHub.tornado_settings = { 'headers': { 'Content-Security-Policy': "frame-ancestors * self host_ip:port"} }
```

change **host_ip** and **port** .It works for me. 

@djknight1

**Comment by taposh** [comment]
@2efPer I tried this and restarted docker, but this did not work for me.

**Comment by nkamsteve** [comment]
@satendrakumar  hello I am a CEO of some on company in Paris. We need to imbedded jupyter on our system. Please can you contact me on LinkedIn :  Steve Nkam Nkam

**Comment by rodriguez-facundo** [comment]
In kubernetes:

```
hub:
  extraConfig:
    hub: |
      c.JupyterHub.tornado_settings = { "headers": { "Content-Security-Policy": "frame-ancestors 'self' http://www.my_address.com:80"}}   <---  or https and any other port
     spawner: >-
       c.Spawner.args = ["--NotebookApp.tornado_settings={\"headers\":{\"Content-Security-Policy\": \"frame-ancestors 'self' http://www.my_address.com:80\"}}" ] 
```

**Comment by soodooo** [comment]
@minrk @rgbkrk @taposh @2efPer @rodriguez-facundo @djknight1 

I solved the problem in this way.
```python
# in jupyterhub_config.py
c.JupyterHub.tornado_settings = {
    'headers': {
        'Content-Security-Policy': "frame-ancestors * ",
    },
}

c.DockerSpawner.extra_create_kwargs.update({'command': '''start-singleuser.sh  
                                            --SingleUserNotebookApp.default_url=/lab  
                                            --SingleUserNotebookApp.tornado_settings="{'headers':{'Content-Security-Policy': ''}}"  
                                            --config=/srv/jupyterhub/jupyter_notebook_config.py'''})
```
###but is anyone who knows that why the config in  `jupyter_notebook_config.py` Do NOT Work?
It looks like my   jupyter_notebook_config.py   is not loaded at all.
```
# in my jupyter_notebook_config.py
c.NotebookApp.tornado_settings = {
    'headers': {
        'Content-Security-Policy': ' ' 
}
```

**Comment by sanjaydatasciencedojo** [comment]
@taposh I guess we have to use same domains for both the sites..

**Comment by AmalVijayan** [comment]
https://github.com/jupyterhub/the-littlest-jupyterhub/issues/312

**Comment by valmack** [comment]
> In kubernetes:
> 
> ```
> hub:
>   extraConfig:
>     hub: |
>       c.JupyterHub.tornado_settings = { "headers": { "Content-Security-Policy": "frame-ancestors 'self' http://www.my_address.com:80"}}   <---  or https and any other port
>      spawner: >-
>        c.Spawner.args = ["--NotebookApp.tornado_settings={\"headers\":{\"Content-Security-Policy\": \"frame-ancestors 'self' http://www.my_address.com:80\"}}" ] 
> ```

I tried this in my config.yaml and tried to upgrade with helm. I'm following the zero-to-jupyterhub kubernetes guide so to upgrade I run
```
helm upgrade $RELEASE jupyterhub/jupyterhub --version=0.8.2 --values config.yaml
```

but I'm still seeing "Blocked by Content Security Policy" in the iframe

**Comment by maluhoss** [comment]
> In kubernetes:
> 
> ```
> hub:
>   extraConfig:
>     hub: |
>       c.JupyterHub.tornado_settings = { "headers": { "Content-Security-Policy": "frame-ancestors 'self' http://www.my_address.com:80"}}   <---  or https and any other port
>      spawner: >-
>        c.Spawner.args = ["--NotebookApp.tornado_settings={\"headers\":{\"Content-Security-Policy\": \"frame-ancestors 'self' http://www.my_address.com:80\"}}" ] 
> ```

^^ this worked for me. Wondering, how do you include multiple iframe URLs (not just www.my_address.com)?

**Comment by hdoupe** [comment]
@maluhoss I believe you can just separate each URL with a space:

`frame-ancestors 'self' http://www.my_address.com:80 http://www.myotheraddress.com`

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors#Syntax

**Comment by maluhoss** [comment]
> @maluhoss I believe you can just separate each URL with a space:
> 
> `frame-ancestors 'self' http://www.my_address.com:80 http://www.myotheraddress.com`
> 
> https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors#Syntax

yeah I tried that initially, but it didnt work :S Still said it couldn't connect.

**Comment by natbusa** [comment]
If you are using nginx as a reverse proxy you can rewrite the policy there without having to fiddle with the jupyterhub machinery

```
proxy_hide_header Content-Security-Policy;  
proxy_set_header Content-Security-Policy "frame-ancestors 'self' http://parenturl.com;";
```

**Comment by EnziinSystem** [comment]
> In kubernetes:
> 
> ```
> hub:
>   extraConfig:
>     hub: |
>       c.JupyterHub.tornado_settings = { "headers": { "Content-Security-Policy": "frame-ancestors 'self' http://www.my_address.com:80"}}   <---  or https and any other port
>      spawner: >-
>        c.Spawner.args = ["--NotebookApp.tornado_settings={\"headers\":{\"Content-Security-Policy\": \"frame-ancestors 'self' http://www.my_address.com:80\"}}" ] 
> ```

Need remove \"

```
hub:
  extraConfig:
    hub: |
      c.JupyterHub.tornado_settings = { 'headers': { 'Content-Security-Policy': 'frame-ancestors self http://domain.com'}}
    spawner: |
      c.Spawner.args = ['--NotebookApp.tornado_settings={"headers":{"Content-Security-Policy": "frame-ancestors * self http://domain.com"}}']
```

**Comment by CITBNU** [comment]
> I found out the easiest way to config, it should work in almost all situation.
> add this to your `jupyterhub_config.py`:
> 
> ```
> c.Spawner.args = ['--NotebookApp.tornado_settings={"headers":{"Content-Security-Policy": "frame-ancestors * self host_ip:port"}}']   
> 
> c.JupyterHub.tornado_settings = { 'headers': { 'Content-Security-Policy': "frame-ancestors * self host_ip:port"} }
> ```
> 
> change **host_ip** and **port** .It works for me.
> 
> @djknight1

THANKS !!!!

**Comment by deekshahegde86** [comment]
> > I found out the easiest way to config, it should work in almost all situation.
> > add this to your `jupyterhub_config.py`:
> > ```
> > c.Spawner.args = ['--NotebookApp.tornado_settings={"headers":{"Content-Security-Policy": "frame-ancestors * self host_ip:port"}}']   
> > 
> > c.JupyterHub.tornado_settings = { 'headers': { 'Content-Security-Policy': "frame-ancestors * self host_ip:port"} }
> > ```
> > 
> > 
> > change **host_ip** and **port** .It works for me.
> > @djknight1
> 
> THANKS !!!!

Can you please share the .py file for reference.

**Comment by ableuler** [comment]
ok, so I guess in the `"frame-ancestors 'self' http://www.my_address.com:80"` from the example `http://www.my_address.com:80` would just be the `https://<renku-host>`, at least whenever we use path based routing?

**Comment by ErqingLi** [comment]
> this

I have the same problem: once the iframe comes to the `user/username/tree?`page, it refuses to connect. 
The solution mentioned by @zcesur works perfectly for me, though I use jupyterhub with DockerSpawner.

To my understanding, the key is to configure `c.NotebookApp.tornado_settings` in `jupyter_notebook_config.py`, which should be within the container we build.  So it won't work if I just configure this in my computer's `etc/.jupyter/` directory.

**Comment by aabdulsalams** [comment]
> If you are using nginx as a reverse proxy you can rewrite the policy there without having to fiddle with the jupyterhub machinery
> 
> ```
> proxy_hide_header Content-Security-Policy;  
> proxy_set_header Content-Security-Policy "frame-ancestors 'self' http://parenturl.com;";
> ```

Thank you, this solved my problem

**Comment by matthewliu0324** [comment]
Update for people that are still struggling, the following setup works for the latest Jupyterhub z2jh version.
`c.JupyterHub.tornado_settings = {'headers': {'Content-Security-Policy': "frame-ancestors *;"}}
c.Spawner.args = ['--NotebookApp.tornado_settings={"headers":{"Content-Security-Policy": "frame-ancestors *;"}}']`
Insert this snippet into your jupyterhub_config.py
